librz/include/rz_arch.h

```diff
@@ -16,12 +20,140 @@ extern "C" {
+	st64 ptr; /* reference to memory */ /* XXX signed? */
+	ut64 val; /* reference to value */ /* XXX signed? */
+	RzAnalysisValue analysis_vals[6]; /* Analyzable values */
+	int ptrsize; /* f.ex: zero extends for 8, 16 or 32 bits only */
+	st64 stackptr; /* stack pointer */
+	int refptr; /* if (0) ptr = "reference" else ptr = "load memory of refptr bytes" */
+	ut64 mmio_address; // mmio address
+	RzAnalysisValue *src[6];
+	RzAnalysisValue *dst;
+	RzList /*<RzAnalysisValue *>*/ *access; /* RzAnalysisValue access information */
+	RzStrBuf esil;
+	RzStrBuf opex;
+	const char *reg; /* destination register */
+	const char *ireg; /* register used for indirect memory computation*/
+	int scale;
+	ut64 disp;
+	RzAnalysisSwitchOp *switch_op;
+	RzAnalysisHint hint;
+	RzAnalysisDataType datatype;
+} RzArchPacket;
+
+/// This needs to be redone, copied for reference from RzAnalysisHint
+typedef struct rz_arch_hint_t {
+	ut64 addr;
+	ut64 ptr;
+	ut64 val; // used to hint jmp rax
+	ut64 jump;
+	ut64 fail;
+	ut64 ret; // hint for function ret values
+	char *arch;
+	char *opcode;
+	char *syntax;
+	char *esil;
+	char *offset;
+	ut32 type;
+	ut64 size;
+	int bits;
+	int new_bits; // change asm.bits after evaluating this instruction
+	int immbase;
+	bool high; // highlight hint
+	int nword;
+	ut64 stackframe;
+} RzArchHint;
+
+typedef enum {
+	RZ_ARCH_DETAIL_ACCESS_UNDEF = 0, ///< Undefined access
+	RZ_ARCH_DETAIL_ACCESS_READ, ///< Read access
+	RZ_ARCH_DETAIL_ACCESS_WRITE, ///< Write access
+} RzArchDetailAccess;
+
+typedef enum {
+	RZ_ARCH_DETAIL_MEMBER_REGISTER = 0, ///< Member is raw bytes
+	RZ_ARCH_DETAIL_MEMBER_UNSIGNED, ///< Member is assembly
+	RZ_ARCH_DETAIL_MEMBER_SIGNED, ///< Member is RzArchPacket
+} RzArchDetailMember;
+
+typedef struct rz_arch_detail_value_t {
+	RzArchDetailMember member;
+	union {
+		size_t register_id;
+		ut64 imm_unsigned;
+		st64 imm_signed;
+	};
+} RzArchDetailValue;
+
+typedef struct rz_arch_detail_t {
+	RzArchDetailValue source[6];
+	RzArchDetailValue destination;
+} RzArchDetail;
+
+typedef void RzArchPluginContext;
+
 typedef struct rz_arch_plugin_t {
-	RZ_DEPRECATE RzAsmPlugin *p_asm; ///< Assembly Plugin
-	RZ_DEPRECATE RzAnalysisPlugin *p_analysis; ///< Analysis Plugin
-	RZ_DEPRECATE RzParsePlugin *p_parse; ///< Parse Plugin
+	RZ_DEPRECATE RzAsmPlugin *p_asm; ///< [Deprecated] Assembly Plugin
+	RZ_DEPRECATE RzAnalysisPlugin *p_analysis; ///< [Deprecated] Analysis Plugin
+	RZ_DEPRECATE RzParsePlugin *p_parse; ///< [Deprecated] Parse Plugin
+
+    bool (*init)(RZ_NONNULL RzConfig *config); ///< Global constructor for the plugin to fill the configuration values.
+    bool (*fini)(); ///< Global destructor for the plugin
+	bool (*can_xcode_in)(RZ_NONNULL RzArchXCodeMember input); ///< Returns true if the plugin can support the given RzArchXCodeMember in input.
+	bool (*can_xcode_out)(RZ_NONNULL RzArchXCodeMember output); ///< Returns true if the plugin can support the given RzArchXCodeMember in ouput.
```

Comment:
An `can_xcode_transform` would be nice. To test if the plugin supports a given `in/out` combination. Because `can_xcode_in() && can_xcode_out() == true` but `context_xcode(in, out)` could be unsupported.

Reply:
i don't think this is an issue because we can define the transform, for example:
`asm > details` could become internally (in rzarch not plugins) `asm > bytes > details`.